Tevinthuku/politico-react-manual-react-config
# Configuring webpack, babel, jest & deployment on heroku

### What does this PR do?

This PR sets up the configuration for tests & web pack settings, proposed js features, code splitting, upon which the app will be built upon.

### Tasks to be accomplished?
- [x] - Install webpack & webpack CLI
- [x] - Set up src directory
- [x] - Setup build script
- [x] - Setup webpack config
- [x] - install babel transpiler
- [x] - set up babel loader to work with webpack
- [x] - install react , react-dom, prop-types
- [x] - setup babel-preset-env & babel/preset-react
- [x] - setup html template file for rendering into
- [x] - install n setup plugin for htmlWebpackPlugin
- [x] - setup different webpack scripts
- [x] - change dev port and set open flag on dev script
- [x] - setup sourcemaps for debugging
- [x] - set up css-loader to inject styles into the app
- [x] - set up hot reloading
- [x] - initialize code-splitting capabilities
- [x] - setup tests env & setup files
- [x] - setup automated code formatting
- [x] - setup eslint with CRA specs
- [x] - add husky and pre-commit git hooks


### How is this to be manually tested?
N/B: Make sure you have [yarn](https://yarnpkg.com/en/docs/install#mac-stable) installed
1. Clone the repo checkout to `ch/setup` branch
2. Run `yarn test` to run the tests
3. Run `yarn dev` to run the dev server or run `yarn dev:hot` to run the dev server with hot reloading
4. Run `yarn build` to build the app in production mode


### Background Context

1. Git hooks
Git hooks have been set up so that we do not commit code whose tests fail and also code that has not been properly formatted. (Formatting is done by prettier)
![image](https://user-images.githubusercontent.com/12128153/57579879-49544580-74ab-11e9-9b33-1fb8d0249e81.png)

<i>gca is an alias for `git commit --amend`</i>

If tests fail then the commit will abort. 

Only well formatted and test passing code will be committed to git.

### Screenshots
N/A
